ufmt/core.py

```diff
@@ -2,118 +2,188 @@
 # Licensed under the MIT license
 
 import logging
-from dataclasses import dataclass
 from functools import partial
 from pathlib import Path
 from typing import List, Optional
+from warnings import warn
 
-from black import (
-    decode_bytes,
-    find_pyproject_toml,
-    format_file_contents,
-    Mode,
-    NothingChanged,
-    parse_pyproject_toml,
-    TargetVersion,
-)
+from black import format_file_contents, NothingChanged
 from moreorless.click import unified_diff
 from trailrunner import Trailrunner
-from usort.config import Config as UsortConfig
-from usort.sorting import usort_string
+from usort import usort
 
 from .config import ufmt_config
+from .types import (
+    BlackConfig,
+    BlackConfigFactory,
+    Encoding,
+    FileContent,
+    PostProcessor,
+    Result,
+    UsortConfig,
+    UsortConfigFactory,
+)
+from .util import make_black_config, read_file, write_file
 
 LOG = logging.getLogger(__name__)
 
 
-@dataclass
-class Result:
-    path: Path
-    changed: bool = False
-    written: bool = False
-    diff: Optional[str] = None
-
-
-def ufmt_string(
+def ufmt_bytes(
     path: Path,
-    content: str,
+    content: FileContent,
+    *,
+    encoding: Encoding = "utf-8",
+    black_config: BlackConfig,
     usort_config: UsortConfig,
-    black_config: Optional[Mode] = None,
-) -> str:
-    content = usort_string(content, usort_config, path)
-    mode = black_config or Mode()
+    post_processor: Optional[PostProcessor] = None,
+) -> FileContent:
+    """
+    Format arbitrary bytes for the given path.
+
+    Requires passing valid config objects for both black and µsort. If the given path
+    represents a type stub (has a ``.pyi`` suffix), the black config object will be
+    updated to set ``is_pyi = True``.
+
+    Note: content will be decoded before passing to black, and re-encoded to bytes
+    again. Be sure to pass a known-valid unicode encoding for the content.
```

Comment:
Why do we need this (the "encoding" param)? usort gives you the encoding as part of the Result, which has a sensible default of utf-8.

I'm disappointed black doesn't have a bytes api.

Reply:
I mostly wanted this as a sanity check, and also mild future proofing in case ufmt wants to do something that needs an encoding before it runs usort.

> I'm disappointed black doesn't have a bytes api.

You and me both, but I think this is mostly a limitation because lib2to3 does not accept bytes. It uses `tokenize.detect_encoding()` but then immediately reopens the file in string mode with the detected encoding.